tests/plugin_tests/chowdsp_plugin_utils_test/GlobalSettingsTest.cpp

```diff
@@ -225,6 +225,53 @@ class GlobalSettingsTest : public TimedUnitTest
         }
     }
 
+    /**
+     * Kick off thread 1, read and write settings to one instance of GlobalPluginSettings
+     * Kick off thread 2 read and write settings to another instance of GlobalPluginSettings
+     * sleep between read/writes to allow for listenerFileChanged callback to be hit, forcing both GlobalPluginSettings instances to read new information from file
+     */
+    void twoInstancesAccessingSameFileTest()
+    {
+        std::atomic<bool> thread1Finished { false };
+        juce::Thread::launch (
+            [&] {
+                chowdsp::GlobalPluginSettings settings;
+                settings.initialise (settingsFile, 1);
+                for (int i = 0; i < 20; ++i)
+                {
+                    auto propId = "thread1_" + std::to_string (i);
+                    chowdsp::GlobalPluginSettings::SettingProperty prop { propId, i };
+                    settings.addProperties ({ prop });
+                    expectEquals (settings.getProperty<int> (prop.first), prop.second.get<int>(), "Property is incorrect within thread 1");
+                    juce::Thread::sleep (50);
+                }
+                thread1Finished = true;
+            });
+
+        std::atomic<bool> thread2Finished { false };
+        juce::Thread::launch (
+            [&] {
+                chowdsp::GlobalPluginSettings settings;
+                settings.initialise (settingsFile, 1);
+                for (int i = 20; i < 40; ++i)
+                {
+                    auto propId = "thread2_" + std::to_string (i);
+                    chowdsp::GlobalPluginSettings::SettingProperty prop { propId, i };
+                    settings.addProperties ({ prop });
+                    expectEquals (settings.getProperty<int> (prop.first), prop.second.get<int>(), "Property is incorrect within thread 2");
+                    juce::Thread::sleep (50);
+                }
+                thread2Finished = true;
+            });
+
+        while (! (thread1Finished && thread2Finished))
+            juce::MessageManager::getInstance()->runDispatchLoopUntil (100);
+
+        chowdsp::GlobalPluginSettings settings;
+        settings.initialise (settingsFile, 1);
+        settings.getSettingsFile().deleteFile();
+    }
```

Comment:
Nice! Just out of curiosity, what happens with the test if the `ScopedLock`'s above are removed? Do the `expectEquals(...)` checks fail, or does the program crash because of some data races?

Reply:
This was one of the things I struggled with writing this test, the current GlobalPluginSettings works the large majority of the time in multiple threads without making mistakes. I was able to get failing cases without locks if I really raised the test numbers, like adding 1000 properties in each thread, instead of the 20 that are currently being added. Even with such large numbers sometimes the test cases would pass without locks.

Let me know if you have any thoughts about better ways to test this. To get consistent failing cases would require such much larger numbers, something like 20 seconds to test and the possibility of running out of memory etc... 

There might be a better approach I haven't thought about as well that could test multiple instances without taking up too much memory or time, it seems like multi threaded issues are living up to their reputation of being hard to test here.

Thanks!

Reply:
Hmm, yeah these kinds of things are always tricky to test... 1000 properties might be a bit much, but I bet 100 or 200 would be okay. Another thing that might be interesting to try is having the test run multiple times when it get's called, like:
```cpp
beginTest ("Two Instances Accessing Same File Test");
static constexpr int numIterations = 100; // or some reasonable number
for (int i = 0; i < numIterations; ++i)
        twoInstancesAccessingSameFileTest();
```
I wouldn't worry too much about how long the tests take to run... it is true that the CI pipelines take a little while at the moment, but the the actual running of the tests is fairly quick relative to the time it takes to build, analyze test coverage data, etc.